render/render.js

```diff
@@ -161,86 +161,215 @@ module.exports = function($window) {
 	}
 
 	//update
-	function updateNodes(parent, old, vnodes, recycling, hooks, nextSibling, ns) {
-		if (old === vnodes || old == null && vnodes == null) return
+	/**
+	 * @param {Element|Fragment} parent - the parent element
+	 * @param {Vnode[] | null} old - the list of vnodes of the last `render()` call for
+	 *                               this part of the tree
+	 * @param {Vnode[] | null} vnodes - as above, but for the current `render()` call.
+	 * @param {boolean} recyclingParent - was the parent vnode or one of its ancestor
+	 *                                    fetched from the recycling pool?
+	 * @param {Function[]} hooks - an accumulator of post-render hooks (oncreate/onupdate)
+	 * @param {Element | null} nextSibling - the next DOM node if we're dealing with a
+	 *                                       fragment that is not the last item in its
+	 *                                       parent
+	 * @param {'svg' | 'math' | String | null} ns) - the current XML namespace, if any
+	 * @returns void
+	 */
+	// This function diffs and patches lists of vnodes, both keyed and unkeyed.
+	//
+	// We will:
+	//
+	// 1. describe its general structure
+	// 2. focus on the diff algorithm optimizations
+	// 3. describe how the recycling pool meshes into this
+	// 4. discuss DOM node operations.
+
+	// ## Overview:
+	//
+	// The updateNodes() function:
+	// - deals with trivial cases
+	// - determines whether the lists are keyed or unkeyed
+	//   (we take advantage of the fact that mixed diff is not supported and settle on the
+	//   keyedness of the first vnode we find)
```

Comment:
Just to clarify, is this actually documented, or is this comment it? (BTW, I was already aware this was the case; I was just wondering.)

Reply:
IIRC it is "undefined behavior" in the sense that it isn't discussed in the docs, intentionally. But see #2046... Previously, two non-null, unkeyed nodes were required to deem a list unkeyed for sure. I'll probably backtrack here for the v1 branch.